homeassistant/components/onkyo/media_player.py

```diff
@@ -233,20 +237,40 @@ async def async_setup_platform(hass, config, async_add_entities, discovery_info=
 
     _LOGGER.debug("Provisioning Onkyo AVR device at %s:%d", host, port)
 
+    active_zones = {}
+
+    def async_onkyo_discover_zones_callback(message):
+        """Receive the power status of the available zones on the AVR."""
+        zone, command, _ = message
+        if command in ["system-power", "power"]:
```

Comment:
Do we really need this? Does it matter if it's a different message? It still means the zone is available, doesn't it?

Reply:
I think we can leave this out. But in that case, I think we can remove the entire `async_onkyo_discover_zones_callback` function and move the lines below into the `else` part of `async_onkyo_update_callback`:
`
zone_entity = OnkyoAVR(avr, f"{name} {zones[zone]}", sources, zone, max_volume)
active_zones[zone] = zone_entity
async_add_entities([zone_entity])
`
What do you think about this?

Reply:
We still need the second condition (`if zone in zones:`), don't we? If that's the case, I think a separate function works better than nested conditions or an `elif`.

Reply:
If no zones are configured manually, the condition `if zone in zones` will always be true anyways.
If the user does configure zones manually and make a typo or so, the discovered zone will not be in `zones` and thus no entity will be created for it. So by removing the condition, a discovered zone will still be added even when it's not in the configuration yaml. What do you think is preferred here?

Reply:
Also, having it in a separate function does mean we have to unpack the message twice.

Reply:
We don't need to unpack twice - we can simply pass the zone to the function :)
As for the correct behavior, I would say whatever the current behavior in the release version is. Since I only use one zone, I don't know what it is.

Reply:
Aah of course, we don't need the entire message in `async_onkyo_discover_zones_callback` but only the zone.
As for the behavior in the current release: it is not possible to specify zones in the configuration yaml there; they are always discovered. This behavior doesn't change when the zones are not configured.

So what is the benefit of adding zones in the configuration? I think the possibility to give it a custom name, like I implemented already. And maybe to only use the specified zones instead of all discovered zones?
My receiver has a zone 2 and a zone 3, but I only use the main and zone 2, so I just don't put zone 3 in the configuration.

Reply:
If that's the case I don't think we need this configuration. We just need to implement `def unique_id` correctly for the zones, and the user will be able to rename and disable from the UI.

Reply:
Yes, I had also been looking into that. The [docs](https://developers.home-assistant.io/docs/entity_registry_index/#unique-id-requirements) state that we should use something like a MAC or serial number, but I couldn't really figure out yet how to do it here. We could maybe implement the unique id in the same way the name is done now, so take the receiver name and add the zone to it, e.g. `media_player.my_receiver_zone_2`. However, users could change it by changing the name in the configuration, so that's not a good solution. Do you have ideas for this?

Reply:
Well, either we try to get the MAC from the IP, or try to get some unique id from the API, or wait till we implement a config flow and then use the config entry id.

Reply:
The receiver does have a unique identifier that can be retrieved through the API. However it is not implemented yet. I actually want to add this feature for device discovery (See my [PR](https://github.com/winterscar/python-eiscp/pull/3) to add discovery in the pyeiscp package). Should we wait for the update there and use that identifier as unique id?

Reply:
Is the identifier only available through discovery?

Reply:
Well, the receiver responds with the identifier when receiving a specific discovery packet. For discovery, I broadcast the discovery packet over the available broadcast network addresses. But it should be possible to send the discovery packet only to the specified host to only retrieve the identifier for this device. To do that, I would need to change a few more things in the pyeiscp package. But for now we can just implement zone discovery in the same way it it done in the current release, without a unique id. The unique id can be added once the pyeiscp package is updated.

Reply:
Yeah, I think that's the right approach. In the meantime, the user won't be able to change zones (which, as you say, is not a regression), and once discovery is back, we'll have a unique id and they'll be able to rename/disable through the UI.